File: src/lib/middleware/rate-limit.ts
import { NextRequest, NextResponse } from 'next/server'

// Simple in-memory rate limiting
// In production, use Redis or similar
const rateLimitStore: Map<string, { count: number; resetAt: number }> = new Map()

export interface RateLimitConfig {
  maxRequests: number
  windowMs: number
}

const DEFAULT_CONFIG: RateLimitConfig = {
  maxRequests: 100,
  windowMs: 60 * 1000 // 1 minute
}

const LOGIN_CONFIG: RateLimitConfig = {
  maxRequests: 5,
  windowMs: 60 * 1000 // 5 attempts per minute
}

const PUBLIC_FORM_CONFIG: RateLimitConfig = {
  maxRequests: 5,
  windowMs: 60 * 60 * 1000 // 5 submissions per hour
}

export function getClientIP(request: NextRequest): string {
  const forwarded = request.headers.get('x-forwarded-for')
  if (forwarded) {
    return forwarded.split(',')[0].trim()
  }
  return request.headers.get('x-real-ip') || 'unknown'
}

export function checkRateLimit(
  key: string,
  config: RateLimitConfig = DEFAULT_CONFIG
): { allowed: boolean; retryAfterSeconds?: number } {
  const now = Date.now()
  const record = rateLimitStore.get(key)
  
  if (!record || now > record.resetAt) {
    // New window
    rateLimitStore.set(key, {
      count: 1,
      resetAt: now + config.windowMs
    })
    return { allowed: true }
  }
  
  if (record.count >= config.maxRequests) {
    const retryAfterSeconds = Math.ceil((record.resetAt - now) / 1000)
    return { allowed: false, retryAfterSeconds }
  }
  
  record.count++
  return { allowed: true }
}

export function rateLimitResponse(retryAfterSeconds: number): NextResponse {
  return NextResponse.json(
    {
      detail: {
        error: 'rate_limit_exceeded',
        message: `Çok fazla istek. ${Math.ceil(retryAfterSeconds / 60)} dakika sonra tekrar deneyin.`,
        retry_after_seconds: retryAfterSeconds
      }
    },
    {
      status: 429,
      headers: {
        'Retry-After': retryAfterSeconds.toString()
      }
    }
  )
}

// Rate limit for login endpoint
export function checkLoginRateLimit(request: NextRequest, email?: string): { allowed: boolean; retryAfterSeconds?: number } {
  const ip = getClientIP(request)
  const key = email ? `login:${ip}:${email}` : `login:${ip}`
  return checkRateLimit(key, LOGIN_CONFIG)
}

// Rate limit for public form submissions
export function checkPublicFormRateLimit(request: NextRequest): { allowed: boolean; retryAfterSeconds?: number } {
  const ip = getClientIP(request)
  const key = `public_form:${ip}`
  return checkRateLimit(key, PUBLIC_FORM_CONFIG)
}

// Cleanup old entries periodically (call this occasionally)
export function cleanupRateLimitStore(): void {
  const now = Date.now()
  for (const [key, record] of rateLimitStore.entries()) {
    if (now > record.resetAt) {
      rateLimitStore.delete(key)
    }
  }
}

// Set interval to cleanup (every 5 minutes)
if (typeof setInterval !== 'undefined') {
  setInterval(cleanupRateLimitStore, 5 * 60 * 1000)
}

File: backend/middleware/rate_limit.py
"""
Rate Limiting Middleware for FastAPI
Simple in-memory implementation - Use Redis for production scaling
"""
from fastapi import Request, HTTPException
from datetime import datetime, timedelta
from typing import Dict, Tuple
import asyncio
import logging
import os

logger = logging.getLogger(__name__)

# Configuration from environment
RATE_LIMIT_LOGIN = int(os.environ.get("RATE_LIMIT_LOGIN", "15"))  # per 15 min (increased for admin friendliness)
RATE_LIMIT_LOGIN_WINDOW = int(os.environ.get("RATE_LIMIT_LOGIN_WINDOW", "900"))  # 15 min in seconds
RATE_LIMIT_REFRESH = int(os.environ.get("RATE_LIMIT_REFRESH", "30"))  # per 15 min
RATE_LIMIT_PUBLIC_FORMS = int(os.environ.get("RATE_LIMIT_PUBLIC_FORMS", "20"))  # per hour
RATE_LIMIT_PUBLIC_FORMS_WINDOW = int(os.environ.get("RATE_LIMIT_PUBLIC_FORMS_WINDOW", "3600"))  # 1 hour

# In-memory storage (use Redis for production)
_rate_limit_store: Dict[str, Dict[str, list]] = {
    "login": {},
    "refresh": {},
    "public_forms": {},
}

# Lock for thread safety
_rate_lock = asyncio.Lock()


async def check_rate_limit(
    key: str,
    limit_type: str,
    max_requests: int,
    window_seconds: int
) -> Tuple[bool, int]:
    """
    Check if the request should be rate limited.
    Returns (is_allowed, retry_after_seconds)
    """
    async with _rate_lock:
        now = datetime.now()
        window_start = now - timedelta(seconds=window_seconds)
        
        if limit_type not in _rate_limit_store:
            _rate_limit_store[limit_type] = {}
        
        store = _rate_limit_store[limit_type]
        
        if key not in store:
            store[key] = []
        
        # Clean old entries
        store[key] = [ts for ts in store[key] if ts > window_start]
        
        # Check limit
        if len(store[key]) >= max_requests:
            # Calculate retry after
            oldest = min(store[key])
            retry_after = int((oldest + timedelta(seconds=window_seconds) - now).total_seconds())
            return False, max(retry_after, 1)
        
        # Add current request
        store[key].append(now)
        return True, 0


async def rate_limit_login(request: Request, email: str = None):
    """Rate limit for login attempts"""
    ip = request.client.host if request.client else "unknown"
    
    # Check IP-based limit
    is_allowed, retry_after = await check_rate_limit(
        f"ip:{ip}",
        "login",
        RATE_LIMIT_LOGIN,
        RATE_LIMIT_LOGIN_WINDOW
    )
    
    if not is_allowed:
        logger.warning(f"Rate limit exceeded for login from IP: {ip}")
        raise HTTPException(
            status_code=429,
            detail={
                "error": "Çok fazla giriş denemesi",
                "message": "Lütfen daha sonra tekrar deneyin.",
                "retry_after_seconds": retry_after
            }
        )
    
    # Also check email-based limit if provided
    if email:
        is_allowed, retry_after = await check_rate_limit(
            f"email:{email}",
            "login",
            RATE_LIMIT_LOGIN,
            RATE_LIMIT_LOGIN_WINDOW
        )
        
        if not is_allowed:
            logger.warning(f"Rate limit exceeded for login email: {email}")
            raise HTTPException(
                status_code=429,
                detail={
                    "error": "Çok fazla giriş denemesi",
                    "message": "Lütfen daha sonra tekrar deneyin.",
                    "retry_after_seconds": retry_after
                }
            )


async def rate_limit_refresh(request: Request):
    """Rate limit for token refresh"""
    ip = request.client.host if request.client else "unknown"
    
    is_allowed, retry_after = await check_rate_limit(
        f"ip:{ip}",
        "refresh",
        RATE_LIMIT_REFRESH,
        RATE_LIMIT_LOGIN_WINDOW
    )
    
    if not is_allowed:
        logger.warning(f"Rate limit exceeded for refresh from IP: {ip}")
        raise HTTPException(
            status_code=429,
            detail={
                "error": "Çok fazla istek",
                "message": "Lütfen daha sonra tekrar deneyin.",
                "retry_after_seconds": retry_after
            }
        )


async def rate_limit_public_form(request: Request):
    """Rate limit for public form submissions"""
    ip = request.client.host if request.client else "unknown"
    
    is_allowed, retry_after = await check_rate_limit(
        f"ip:{ip}",
        "public_forms",
        RATE_LIMIT_PUBLIC_FORMS,
        RATE_LIMIT_PUBLIC_FORMS_WINDOW
    )
    
    if not is_allowed:
        logger.warning(f"Rate limit exceeded for public form from IP: {ip}")
        raise HTTPException(
            status_code=429,
            detail={
                "error": "Çok fazla form gönderimi",
                "message": "Lütfen bir süre bekleyip tekrar deneyin.",
                "retry_after_seconds": retry_after
            }
        )


def clear_rate_limit_store():
    """Clear all rate limit data - useful for testing"""
    global _rate_limit_store
    _rate_limit_store = {
        "login": {},
        "refresh": {},
        "public_forms": {},
    }
